core/src/stored/append.cc

```diff
@@ -111,6 +118,164 @@ static bool SaveFullyProcessedFilesAttributes(
   return false;
 }
 
+class MessageHandler {
+ public:
+  using signal = int;
+
+  struct message {
+    std::size_t size;
+    PoolMem data;
+  };
+
+  struct error {
+    enum class type
+    {
+      HARDEOF,
+      // both ERROR and SOCKET_ERROR are taken by windows.h
+      INTERNAL_ERROR
+    } type;
+
+    std::string msg;
+  };
+
+  using result_type = std::variant<signal, message, error>;
+
+  MessageHandler(BareosSocket* fd) : fd{fd}, receive_thread{enlist, this} {}
+
+  std::optional<result_type> get_msg()
+  {
+    if (cache.size() == 0) {
+      std::unique_lock l(mut);
+      not_empty.wait(l, [this] { return results.size() > 0 || finished; });
+
+      if (results.size() == 0) { return std::nullopt; }
+
+      std::swap(cache, results);
+    }
+
+    ASSERT(cache.size() > 0);
+
+    result_type t = std::move(cache.front());
+    cache.pop_front();
+
+    std::unique_lock l(mut);
+    bytes_out += sizeof(t);
+    if (auto* content = std::get_if<message>(&t)) {
+      bytes_out += content->size;
+    }
+    element_removed.notify_one();
+    return t;
+  }
+
+  BareosSocket* close_and_get()
+  {
+    end.store(true);
+    receive_thread.join();
+    return fd;
+  }
+
+ private:
+  bool error_while_reading{false};
+  std::atomic<bool> end{false};
+
+  BareosSocket* fd;
+  // channel::out<result_type> chan;
+
+  std::mutex mut{};
+  bool finished{false};
+  std::size_t bytes_out{0};
+  std::condition_variable not_empty{};
+  std::condition_variable element_removed{};
+  std::deque<result_type> results{};
+
+  std::deque<result_type> cache;
```

Comment:
```suggestion
  std::deque<result_type> cache{};
```
(or remove the `{}` from all members that are guaranteed to be constructed somehow)